spring-xd-dirt/src/test/java/org/springframework/xd/dirt/batch/tasklet/JobLaunchingTaskletTests.java

```diff
@@ -0,0 +1,377 @@
+import org.springframework.batch.core.scope.context.ChunkContext;
+import org.springframework.batch.core.scope.context.StepContext;
+import org.springframework.batch.repeat.RepeatStatus;
+import org.springframework.core.task.SimpleAsyncTaskExecutor;
+import org.springframework.core.task.TaskExecutor;
+import org.springframework.integration.amqp.channel.PublishSubscribeAmqpChannel;
+import org.springframework.integration.channel.PublishSubscribeChannel;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.xd.dirt.integration.bus.MessageBus;
+import org.springframework.xd.dirt.stream.JobDefinition;
+import org.springframework.xd.dirt.stream.JobDefinitionRepository;
+import org.springframework.xd.dirt.stream.NoSuchDefinitionException;
+import org.springframework.xd.dirt.stream.NotDeployedException;
+import org.springframework.xd.store.DomainRepository;
+
+/**
+ * @author Michael Minella
+ */
+public class JobLaunchingTaskletTests {
+
+	private JobLaunchingTasklet tasklet;
+
+	@Mock
+	private MessageBus bus;
+
+	@Mock
+	private JobDefinitionRepository jobDefinitionRepository;
+
+	@Mock
+	private DomainRepository<JobDefinition, String> instanceRepository;
+
+	@Mock
+	private MessageChannel launchingChannel;
+
+	@Spy
+	private PublishSubscribeChannel listeningChannel;
+
+	private TaskExecutor taskExecutor = new SimpleAsyncTaskExecutor();
+
+	@Before
+	public void setUp() {
+		this.listeningChannel = new PublishSubscribeChannel(new SimpleAsyncTaskExecutor());
+		MockitoAnnotations.initMocks(this);
+
+		this.tasklet = new JobLaunchingTasklet(bus, jobDefinitionRepository, instanceRepository, "foo", launchingChannel, listeningChannel);
+	}
+
+	@Test
+	public void testInitialLaunch() throws Exception {
+		JobInstance jobInstance = new JobInstance(3l, "masterFoo");
+		JobExecution jobExecution = new JobExecution(5l);
+		jobExecution.setJobInstance(jobInstance);
+		StepExecution stepExecution = new StepExecution("masterFoo", jobExecution, 7l);
+		final StepContribution stepContribution = new StepContribution(stepExecution);
+		final ChunkContext chunkContext = new ChunkContext(new StepContext(stepExecution));
+
+		JobDefinition jobDefinition = new JobDefinition("foo", "foo");
+		when(jobDefinitionRepository.findOne("foo")).thenReturn(jobDefinition);
+		when(instanceRepository.findOne("foo")).thenReturn(jobDefinition);
+		doReturn(true).when(launchingChannel).send(any(Message.class));
+
+		FutureTask<RepeatStatus> result = new FutureTask<RepeatStatus>(new Callable<RepeatStatus>() {
+			@Override
+			public RepeatStatus call() throws Exception {
+				return tasklet.execute(stepContribution, chunkContext);
+			}
+		});
+
+		taskExecutor.execute(result);
+
+		JobParameters jobParameters = new JobParametersBuilder().addString("xd_orchestration_id", "3").toJobParameters();
```

Comment:
Wonder if we should be using the constant `XD_ORCHESTRATION_ID` instead of plain strings.